system-parachains/encointer/src/migrations_fix.rs

```diff
@@ -17,173 +17,131 @@
 
-pub mod balances {
+pub mod collator_selection_init {
 	use frame_support::traits::OnRuntimeUpgrade;
-	use pallet_balances::*;
 	#[cfg(feature = "try-runtime")]
 	use sp_runtime::TryRuntimeError;
 
 	/// The log target.
-	const TARGET: &'static str = "runtime::fix::balances::migration";
-	pub mod v1 {
+	const TARGET: &'static str = "runtime::fix::collator_selection_init";
+	pub mod v0 {
 		use super::*;
-		use frame_support::pallet_prelude::*;
-		pub struct BruteForceToV1<T>(sp_std::marker::PhantomData<T>);
-
-		impl<T: Config> OnRuntimeUpgrade for BruteForceToV1<T> {
+		use crate::SessionKeys;
+		use codec::EncodeLike;
+		use frame_support::{pallet_prelude::*, traits::Currency};
+		use hex_literal::hex;
+		use log::info;
+		use parachains_common::impls::BalanceOf;
+		use sp_core::{crypto::key_types::AURA, sr25519};
+		use sp_std::{vec, vec::Vec};
+
+		const INVULNERABLE_A: [u8; 32] =
+			hex!("5e962096da68302d5c47fce0178d72fab503c4f00a3f1df64856748f0d9dd51e");
+		const INVULNERABLE_B: [u8; 32] =
+			hex!("0cecb8d1c2c744ca4c5cea57f5d6c40238f4dad17afa213672b8b7d43b80a659");
+		const INVULNERABLE_C: [u8; 32] =
+			hex!("ca1951a3c4e100fb5a899e7bae3ea124491930a72000c5e4b2775fea27ecf05d");
+		const INVULNERABLE_D: [u8; 32] =
+			hex!("484b443bd95068b860c92b0f66487b78f58234eca0f88e2adbe80bae4807b809");
+		const INVULNERABLE_E: [u8; 32] =
+			hex!("6c642fb4b571a5685a869cd291fafd575be47a918b231ba28165e5c0cd0cfa15");
+
+		pub struct InitInvulnerables<T>(sp_std::marker::PhantomData<T>);
+		impl<T> OnRuntimeUpgrade for InitInvulnerables<T>
+		where
+			T: frame_system::Config
+				+ pallet_collator_selection::Config
+				+ pallet_session::Config
+				+ pallet_balances::Config,
+			<T as frame_system::Config>::AccountId: From<[u8; 32]>,
+			<T as pallet_session::Config>::ValidatorId: From<[u8; 32]>,
+			<T as pallet_session::Config>::Keys: From<SessionKeys>,
+			<T as pallet_balances::Config>::Balance: From<u128>,
+			<T as pallet_balances::Config>::Balance: EncodeLike<
+				<<T as pallet_collator_selection::Config>::Currency as Currency<
+					<T as frame_system::Config>::AccountId,
+				>>::Balance,
+			>,
+		{
 			#[cfg(feature = "try-runtime")]
 			fn pre_upgrade() -> Result<sp_std::vec::Vec<u8>, TryRuntimeError> {
-				Ok(().encode())
+				let invulnerables_len = pallet_collator_selection::Invulnerables::<T>::get().len();
+				Ok((invulnerables_len as u32).encode())
 			}
 
 			fn on_runtime_upgrade() -> Weight {
-				let onchain_version = Pallet::<T>::on_chain_storage_version();
-				if onchain_version >= 1 {
-					log::warn!(
-						target: TARGET,
-						"skipping bruteforce to v1 migration: executed on wrong storage version."
-					);
-					return T::DbWeight::get().reads(1)
+				let invulnerables_len = pallet_collator_selection::Invulnerables::<T>::get().len();
+				if invulnerables_len > 0 {
+					info!(target: TARGET, "no need to initialize invulnerables");
+					return T::DbWeight::get().reads_writes(0, 0)
```

Comment:
It had to read `Invulnerables`
```suggestion
					return T::DbWeight::get().reads_writes(1, 0)
```